Sources/NIOFileSystem/FileSystem.swift

```diff
@@ -917,72 +922,228 @@ extension FileSystem {
+                        message: """
+                            Can't copy '\(source.path)' of type '\(source.type)'; only regular \
+                            files, symbolic links and directories can be copied.
+                            """,
+                        cause: nil,
+                        location: .here()
+                    )
+
+                    try await shouldProceedAfterError(source, error)
                 }
             }
+        }
+    }
 
-            return directoriesToCopy
+    /// Copies the directory from `sourcePath` to `destinationPath`.
+    private func copyDirectory(
+        from sourcePath: FilePath,
+        to destinationPath: FilePath,
+        strategy copyStrategy: CopyStrategy,
+        shouldProceedAfterError: @escaping @Sendable (
+            _ entry: DirectoryEntry,
+            _ error: Error
+        ) async throws -> Void,
+        shouldCopyItem: @escaping @Sendable (
+            _ source: DirectoryEntry,
+            _ destination: FilePath
+        ) async -> Bool
+    ) async throws {
+        switch copyStrategy.wrapped {
+        case .sequential:
+            return try await self.copyDirectorySequential(
+                from: sourcePath,
+                to: destinationPath,
+                shouldProceedAfterError: shouldProceedAfterError,
+                shouldCopyItem: shouldCopyItem
+            )
+        case let .parallel(maxDescriptors):
+            // Note that maxDescriptors was validated on construction of CopyStrategy.
+            // See notes on CopyStrategy about assumptions on descriptor use.
+            // For now we take the worst case peak for every operation, which is 2 descriptors,
+            // this keeps the downstream limiting code simple
+            // We do not preclude the use of more granular limiting in future (e.g. a directory
+            // scan only requires 1), for now we just drop any excess remainder entirely.
+            let limitValue = maxDescriptors / 2
+            return try await self.copyDirectoryParallel(
+                from: sourcePath,
+                to: destinationPath,
+                maxConcurrentOperations: limitValue,
+                shouldProceedAfterError: shouldProceedAfterError,
+                shouldCopyItem: shouldCopyItem
+            )
         }
+    }
 
-        for entry in directoriesToCopy {
-            if await shouldCopyFile(entry.from, entry.to) {
-                try await self.copyDirectory(
-                    from: entry.from,
-                    to: entry.to,
-                    shouldProceedAfterError: shouldProceedAfterError,
-                    shouldCopyFile: shouldCopyFile
+    /// Building block of the parallel directory copy implementation
+    /// Each invovation of this is allowed to consume two file descriptors,
+    /// any further work (if any) should be sent to `yield` for future processing
+    @Sendable func copySelfAndEnqueueChildren(
```

Comment:
We should never have to mark methods as `@Sendable`. The compiler should be able to infer that 
```suggestion
    func copySelfAndEnqueueChildren(
```

Reply:
Ah yes, this was previously a closure and I failed to remove that when I pulled it out to an instance method thanks